### ISSUE TITLE
consul connect: add tags/metadata to proxy services

### ISSUE THREAD

**Issue by roidelapluie** [hashicorp/consul#4300]
#### Feature Description

Currently connect proxies comes as services in consul but do not have tags and metadata.

It would be useful to be able to add tags/metadata to them and/or let them inherit tags and metadata from their service.

#### Use Case(s)

Many uses cases as it would enable us to discover the connect proxies and query them using the traditional api calls.

**Comment by mitchellh** [comment]
Proxies are just service entries so they fully support tags and metadata. You are perhaps proposing supporting custom tags/metadata for the _managed_ proxy instances, which is probably a good idea. An idea then is to expose the same tags/metadata fields in the `connect { proxy { } }` stanzas within a service definition. This would be fairly simple I think. Thoughts @banks?

For unmanaged proxies, they are just normal service definitions with "kind" set to "connect-proxy" and a few other fields. They fully support tags/metadata as-is.

And for your target use case of discovering connect proxies, we have a number of API endpoints for this already, since we differentiate them in the catalog. You can search by proxy itself or by the application being proxied. You can also just use the non-proxy API endpoints, filter by tag/metadata, and then filter locally (not the API itself) by the "kind" attribute.

**Comment by banks** [comment]
Yeah thanks for the report.

@roidelapluie it makes sense to me to have managed proxies inherit the tags and meta from the service since it's only a transport detail that there is a proxy at all - the routing decisions you make based on tag or meta shoud work the same with or without a proxy.

@mitchellh we could add overrides to the managed proxy config like you suggest but I feel like it's more logical to just mirror them for now unless people come up with legitimate use-cases for adding specific and different meta to a proxy than the service behind it? We already have a known issue where we need to copy the target service's _checks_ in some way and this feels the same - have discovery work the same for the proxy as it does for the service.

**Comment by mitchellh** [comment]
@banks I believe mirroring may not be the right approach _generally_.... but to discuss:

Proxies today represent a service by _name_ via the `ProxyDestination` field in the service definition. They do not proxy by _id_. So a proxy represents ANY service with that name. This was done consciously during the dev cycle early on so that load-balancer type proxies could be supported. Therefore, filtering by tag/meta may not work as expected.

But, most proxies do exactly represent one service and the ones we know that do for sure are the ones that are specified inline with a service definition (managed proxies) via `connect { proxy { ... } }` configurations. 

**So we _could_ support a `mirror_tags_meta` (hypothetical) field in that block and support it for managed proxies, but I don't think its necessarily right at the general `agent/consul` RPC layer.** And that would probably go very far to solve this issue, because you can specify tags/meta as you see fit on unmanaged proxies (they're just service definitions). 

Longer term, we probably need to add to the data model a `ProxyServiceId` field to the service definition to support proxies that are sidecars (almost all of them, 99%) and so we can do more clever health checking and so on. But I think supporting general "name" is good, too.

So, its more complex than I think we initially thought, but wanted to get all that detail out in the GH thread.

**Comment by banks** [comment]
Yeah @mitchellh `ProxyServiceId` is where was imagining this would go in practice - that would be optional and only specified for a side car style proxy that fronts a single instance of some service.

That then allows us to "inherit" implicitly all the things like tags, meta, check status etc. during discovery.

I'm not sure why we'd need explicit option (like `mirror_tags_meta`) for this -- it would only be used for managed proxies and I don't see a case where you'd want different behaviour (concretely in agent/state `AddProxy` where we define the service to register).

For unmanaged proxies it's still possible to support them fronting multiple instances with all of these features since the user registers the service directly.
